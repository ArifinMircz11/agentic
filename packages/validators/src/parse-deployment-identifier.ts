import { HttpError } from '@agentic/platform-core'

import type {
  ParsedDeploymentIdentifier,
  ParseIdentifierOptions
} from './types'
import { parseToolIdentifier } from './parse-tool-identifier'
import { coerceIdentifier } from './utils'

const deploymentIdentifierImplicitRe =
  /^@([a-z0-9-]{1,256})\/([a-z0-9-]{1,256})$/

const deploymentIdentifierHashRe =
  /^@([a-z0-9-]{1,256})\/([a-z0-9-]{1,256})@([a-z0-9]{8})$/

const deploymentIdentifierVersionRe =
  /^@([a-z0-9-]{1,256})\/([a-z0-9-]{1,256})@([\d.a-z-@]+)$/

export function parseDeploymentIdentifier(
  identifier?: string,
  { strict = true, errorStatusCode = 400 }: ParseIdentifierOptions = {}
): ParsedDeploymentIdentifier {
  const inputIdentifier = identifier

  if (!strict) {
    try {
      return parseToolIdentifier(identifier, {
        strict,
        errorStatusCode
      })
    } catch {
      // ignore
    }
  }

  if (!strict) {
    identifier = coerceIdentifier(identifier)
  }

  if (!identifier?.length) {
    throw new HttpError({
      statusCode: errorStatusCode,
      message: `Invalid deployment identifier "${inputIdentifier}"`
    })
  }

  const iMatch = identifier.match(deploymentIdentifierImplicitRe)

  if (iMatch) {
    return {
      projectIdentifier: `@${iMatch[1]!}/${iMatch[2]!}`,
      projectNamespace: iMatch[1]!,
      projectName: iMatch[2]!,
      deploymentIdentifier: `@${iMatch[1]!}/${iMatch[2]!}@latest`,
      deploymentVersion: 'latest'
    }
  }

  const hMatch = identifier.match(deploymentIdentifierHashRe)

  if (hMatch) {
    return {
      projectIdentifier: `@${hMatch[1]!}/${hMatch[2]!}`,
      projectNamespace: hMatch[1]!,
      projectName: hMatch[2]!,
      deploymentIdentifier: `@${hMatch[1]!}/${hMatch[2]!}@${hMatch[3]!}`,
      deploymentHash: hMatch[3]!
    }
  }

  const vMatch = identifier.match(deploymentIdentifierVersionRe)

  if (vMatch) {
    return {
      projectIdentifier: `@${vMatch[1]!}/${vMatch[2]!}`,
      projectNamespace: vMatch[1]!,
      projectName: vMatch[2]!,
      deploymentIdentifier: `@${vMatch[1]!}/${vMatch[2]!}@${vMatch[3]!}`,
      deploymentVersion: vMatch[3]!
    }
  }

  throw new HttpError({
    statusCode: errorStatusCode,
    message: `Invalid deployment identifier "${inputIdentifier}"`
  })
}
