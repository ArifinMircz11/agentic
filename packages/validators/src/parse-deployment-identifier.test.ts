import { describe, expect, test } from 'vitest'

import { parseDeploymentIdentifier } from './parse-deployment-identifier'
import {
  isValidDeploymentHash,
  isValidDeploymentIdentifier,
  isValidNamespace,
  isValidProjectIdentifier,
  isValidProjectSlug
} from './validators'

function success(...args: Parameters<typeof parseDeploymentIdentifier>) {
  const result = parseDeploymentIdentifier(...args)
  expect(result).toBeTruthy()
  expect(result!.projectIdentifier).toBeTruthy()
  expect(result!.projectNamespace).toBeTruthy()
  expect(result!.projectSlug).toBeTruthy()
  expect(result!.deploymentIdentifier).toBeTruthy()
  expect(result!.deploymentVersion || result!.deploymentHash).toBeTruthy()
  expect(isValidProjectIdentifier(result!.projectIdentifier)).toBe(true)
  expect(isValidProjectSlug(result!.projectSlug)).toBe(true)
  expect(isValidNamespace(result!.projectNamespace)).toBe(true)
  expect(isValidDeploymentIdentifier(result!.deploymentIdentifier)).toBe(true)

  if (result!.deploymentHash) {
    expect(isValidDeploymentHash(result!.deploymentHash)).toBe(true)
  }

  expect(result).toMatchSnapshot()
}

function error(...args: Parameters<typeof parseDeploymentIdentifier>) {
  expect(() => parseDeploymentIdentifier(...args)).throws()
}

describe('parseDeploymentIdentifier', () => {
  test('strict mode valid', () => {
    success('@username/foo-bar')
    success('@username/foo-bar@01234567')
    success('@username/foo-bar@latest')
    success('@username/foo-bar@dev')
    success('@username/foo-bar@1.0.0')
    success('@username/foo-bar@abc123lz')
    success('@username/foobar123-yo@01234567')

    success('@u/foo-bar')
    success('@a/foo-bar')
    success('@foo/foobar123-yo')
    success('@foo/foobar123-yo')
    success('@foo/bar')
  })

  test('strict mode invalid', () => {
    error()
    error('')
    error('foo')
    error('foo/bar')
    error('foo/bar/baz')
    error('foo/bar/baz/qux')
    error('@foo/bar/baz/qux')
    error('@foo/bar/baz/qux/quux')
    error('@foo/bar/baz/qux/quux/corge')
    error('@foo/bar/baz/qux/quux/corge/grault')
    error('@foo/bar/baz/qux/quux/corge/grault/garply')
    error('@foo/bar/baz/')
    error('@foo')
    error('@foo/bar/')

    error('@foo-bar@01234567')
    error('@%/foo-bar@01234567')
    error('@user/foo^bar@01234567')
    error('@user@foo^bar@01234567')
    error('@username/Foo-Bar@01234567')
    error('username/foo-bar')
    error('@Username/foo-bar')
    error('@username/Foo-bar')
    error('@username/foo-bar/')
    error('@username/foo_bar')

    error('@foo_bar@latest')

    error('//@username/foo-bar@01234567')
    error('https://@username/foo-bar@01234567')
    error('https://example.com/@username/foo-bar')
  })

  test('non-strict mode valid', () => {
    success('https://gateway.agentic.so/@username/foo-bar', { strict: false })
    success('https://gateway.agentic.so/@username/foo-bar@latest', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@01234567', {
      strict: false
    })
    success('/@username/foo-bar@01234567', { strict: false })
    success('@username/foo-bar@01234567/', { strict: false })

    success('/@username/foo-bar@01234567/', { strict: false })
    success('/@username/foo-bar@latest/', { strict: false })
    success('/@username/foo-bar@dev/', { strict: false })
    success('/@username/foo-bar@1.0.0/', { strict: false })
    success('/@username/foo-bar/', { strict: false })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@latest', {
      strict: false
    })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo', {
      strict: false
    })
    success('/@username/foo-bar@01234567/foo', { strict: false })
    success('@username/foo-bar@01234567/foo/', { strict: false })
    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })

    success('/@username/foo-bar@01234567/foo/', { strict: false })
    success('/@username/foo-bar@latest/foo/', { strict: false })
    success('/@username/foo-bar@dev/foo/', { strict: false })
    success('/@username/foo-bar@1.0.0/foo/', { strict: false })
    success('/@username/foo-bar/foo/', { strict: false })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@latest/foo/', {
      strict: false
    })
  })

  test('non-strict mode invalid', () => {
    error(undefined, { strict: false })
    error('', { strict: false })
    error('https://gateway.agentic.so', { strict: false })
    error('//gateway.agentic.so', { strict: false })
    error('https://gateway.agentic.so/@username', { strict: false })
    error('https://gateway.agentic.so/@username/', { strict: false })
    error('https://gateway.agentic.so/call/@username/foo-bar@latest/tool', {
      strict: false
    })
    error('/call/@username/foo-bar/foo', { strict: false })
    error('//@username/foo-bar/foo', { strict: false })
    error('@username/foo-bar/foo//', { strict: false })

    error('@username/foo-bar/fooðŸ˜€', { strict: false })
    error('@username/Foo-Bar@dev/foo/', { strict: false })

    error('@username/Foo', { strict: false })
    error('@username/fooðŸ˜€', { strict: false })
  })
})
