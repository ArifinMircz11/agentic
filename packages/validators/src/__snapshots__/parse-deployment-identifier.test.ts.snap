// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseDeploymentIdentifier > non-strict mode valid 1`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 2`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 6`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 7`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 8`] = `
{
  "deploymentIdentifier": "@username/foo-bar@dev",
  "deploymentVersion": "dev",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 9`] = `
{
  "deploymentIdentifier": "@username/foo-bar@1.0.0",
  "deploymentVersion": "1.0.0",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 10`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 11`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 12`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 13`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 14`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 15`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 16`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 17`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 18`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 19`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 20`] = `
{
  "deploymentIdentifier": "@username/foo-bar@dev",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 21`] = `
{
  "deploymentIdentifier": "@username/foo-bar@1.0.0",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 22`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 23`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 24`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > non-strict mode valid 25`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
  "toolName": "foo",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 1`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foo-bar@01234567",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 3`] = `
{
  "deploymentIdentifier": "@username/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 4`] = `
{
  "deploymentIdentifier": "@username/foo-bar@dev",
  "deploymentVersion": "dev",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 5`] = `
{
  "deploymentIdentifier": "@username/foo-bar@1.0.0",
  "deploymentVersion": "1.0.0",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 6`] = `
{
  "deploymentHash": "abc123lz",
  "deploymentIdentifier": "@username/foo-bar@abc123lz",
  "projectIdentifier": "@username/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 7`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "@username/foobar123-yo@01234567",
  "projectIdentifier": "@username/foobar123-yo",
  "projectName": "foobar123-yo",
  "projectNamespace": "username",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 8`] = `
{
  "deploymentIdentifier": "@u/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@u/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "u",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 9`] = `
{
  "deploymentIdentifier": "@a/foo-bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@a/foo-bar",
  "projectName": "foo-bar",
  "projectNamespace": "a",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 10`] = `
{
  "deploymentIdentifier": "@foo/foobar123-yo@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@foo/foobar123-yo",
  "projectName": "foobar123-yo",
  "projectNamespace": "foo",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 11`] = `
{
  "deploymentIdentifier": "@foo/foobar123-yo@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@foo/foobar123-yo",
  "projectName": "foobar123-yo",
  "projectNamespace": "foo",
}
`;

exports[`parseDeploymentIdentifier > strict mode valid 12`] = `
{
  "deploymentIdentifier": "@foo/bar@latest",
  "deploymentVersion": "latest",
  "projectIdentifier": "@foo/bar",
  "projectName": "bar",
  "projectNamespace": "foo",
}
`;
